auth_jwt/models/auth_jwt_validator.py

```diff
@@ -0,0 +1,173 @@
+        return validator
+
+    @tools.ormcache("self.public_key_jwk_uri", "kid")
+    def _get_key(self, kid):
+        r = requests.get(self.public_key_jwk_uri)
+        r.raise_for_status()
+        response = r.json()
+        for key in response["keys"]:
+            if key["kid"] == kid:
+                return key
+        return {}
+
+    def _decode(self, token):
+        """Validate and decode a JWT token, return the payload."""
+        if self.signature_type == "secret":
+            key = self.secret_key
+            algorithm = self.secret_algorithm
+        else:
+            try:
+                header = jwt.get_unverified_header(token)
+            except jwt.JWTError as e:
+                _logger.info("Invalid token: %s", e)
+                raise AccessDeniedInvalidToken()
+            key = self._get_key(header.get("kid"))
+            algorithm = self.public_key_algorithm
+        try:
+            payload = jwt.decode(
+                token,
+                key=key,
+                algorithms=[algorithm],
+                options=dict(
+                    require_exp=True, verify_exp=True, verify_aud=True, verify_iss=True
+                ),
+                audience=self.audience,
+                issuer=self.issuer,
+            )
+        except jwt.JWTError as e:
+            _logger.info("Invalid token: %s", e)
+            raise AccessDeniedInvalidToken()
+        return payload
+
+    def _get_uid(self, payload):
+        # override for additional strategies
+        if self.user_id_strategy == "static":
+            return self.static_user_id.id
+
+    def _get_and_check_uid(self, payload):
+        uid = self._get_uid(payload)
+        if not uid:
+            _logger.error("_get_uid did not return a user id")
+            raise AccessDenied()
+        return uid
+
+    def _get_partner_id(self, payload):
+        # override for additional strategies
+        if self.partner_strategy == "email":
+            email = payload.get("email")
+            if not email:
+                _logger.debug("JWT payload does not have an email claim")
+                return
+            partner = self.env["res.partner"].search([("email", "=", email)])
+            if len(partner) != 1:
+                _logger.debug("%d partners found for email %s", len(partner), email)
+                return
+            return partner.id
+
+    def _get_and_check_partner_id(self, payload):
+        partner_id = self._get_partner_id(payload)
+        if not partner_id and self.partner_required:
+            raise AccessDenied()
+        return partner_id
+
+    def _register_hook(self):
+        res = super()._register_hook()
+        self.search([])._register_auth_method()
+        return res
+
+    def _register_auth_method(self):
+        IrHttp = self.env["ir.http"]
+        for rec in self:
+            setattr(
+                IrHttp.__class__,
+                f"_auth_method_jwt_{rec.name}",
```

Comment:
@sbidoul We should declare a constraint on ` name` to be sure that it doesn't contain invalid characters when used as part of a method name.

Reply:
Yes, absolutely, that is on my todo list.